ledger-service/http-json/src/main/scala/com/digitalasset/http/EndpointsCompanion.scala

```diff
@@ -55,22 +64,134 @@ object EndpointsCompanion {
       case NonFatal(t) => ServerError(t.description)
     }
   }
-
+  private type ET[A] = EitherT[Future, Unauthorized, A]
   trait ParsePayload[A] {
-    def parsePayload(jwt: DecodedJwt[String]): Unauthorized \/ A
+    def parsePayload(
+        jwt: DecodedJwt[String]
+    ): Error \/ A
+  }
+
+  trait CreateFromUserToken[A] {
+    def apply(
+        jwt: DecodedJwt[String],
+        listUserRights: UserId => Future[Vector[UserRight]],
+        getLedgerId: () => Future[String],
+    ): ET[A]
+  }
+
+  object CreateFromUserToken {
+
+    import com.daml.http.util.FutureUtil.either
+
+    trait FromUser[A, B] {
+      def apply(userId: String, actAs: List[String], readAs: List[String], ledgerId: String): A \/ B
+    }
+
+    private def transformUserTokenTo[B](
+        jwt: DecodedJwt[String],
+        listUserRights: UserId => Future[Vector[UserRight]],
+        getLedgerId: () => Future[String],
+    )(
+        fromUser: FromUser[Unauthorized, B]
+    )(implicit
+        mf: Monad[Future]
+    ): EitherT[Future, Unauthorized, B] = {
+      import scalaz.syntax.std.option._
+      import scalaz.syntax.std.either._
+      import spray.json._
+      for {
+        jsValue <- either(
+          Try(jwt.payload.parseJson).toEither.disjunction
+            .leftMap(it => Unauthorized(it.getMessage))
+        )
+        decodedToken <- either(
+          readStandardTokenPayload(
+            jsValue
+          ) \/> Unauthorized("Invalid token supplied")
+        )
+
+        userId <- either(
+          (decodedToken.applicationId \/> Unauthorized(
+            "This user token contains no applicationId/userId"
+          )).flatMap(UserId.fromString(_).left.map(Unauthorized).disjunction)
+        )
+        rights <- EitherT.rightT(listUserRights(userId))
+
+        actAs = rights.collect { case CanActAs(party) =>
+          party
+        }
+        readAs = rights.collect { case CanReadAs(party) =>
+          party
+        }
+        ledgerId <- EitherT.rightT(getLedgerId())
+        res <- either(fromUser(userId, actAs.toList, readAs.toList, ledgerId))
+      } yield res
+    }
+
+    private[http] implicit def jwtWritePayloadFromUserToken(implicit
+        mf: Monad[Future]
+    ): CreateFromUserToken[JwtWritePayload] =
+      (
+          jwt,
+          listUserRights,
+          getLedgerId,
+      ) =>
+        transformUserTokenTo(jwt, listUserRights, getLedgerId)((userId, actAs, readAs, ledgerId) =>
```

Comment:
if you pull the call at the end out of `transformUserTokenTo` you also avoid the somewhat confusing nested callbacks/lambdas here.